samcli/commands/local/lib/sam_function_code_provider.py

```diff
@@ -0,0 +1,176 @@
+"""
+Class that provides codeuri from a given Function definition
+"""
+
+import logging
+import os
+import six
+
+LOG = logging.getLogger(__name__)
+
+
+class SamFunctionCodeProvider(object):
+    """
+    Lambda Function Code provider
+
+        Parameters
+        ----------
+        function_name str
+            FunctionId
+        resource_properties dict
+            Function dictionary
+        function_type str
+            Either Serverless::Function or Lambda::Function
+    """
+
+    _SERVERLESS_FUNCTION = "AWS::Serverless::Function"
+    _LAMBDA_FUNCTION = "AWS::Lambda::Function"
+    _DEFAULT_CODEURI = "."
+
+    def __init__(self, function_name, resource_properties, function_type):
+        """ Store Function info
+        This Class is a Provider for the codeuri property hold the path to an existing file/folder
+        or a pointer to a folder used to store InlineCode/ZipFile string property (Inline Lambdas)
+        It also provide a context manager function to dump the code inside a cache path and remove it
+        after invocation
+        """
+        self.function_name = function_name
+        self.handler = resource_properties.get('Handler')
+        self.runtime = resource_properties.get('Runtime')
+        self.index = self._setup_postfix(self.runtime)
+
+        self.inlinecode = self._extract_inline_code(resource_properties, function_type)
+        if self.inlinecode:
+            self.codeuri = './.aws-sam/inline/{}/'.format(function_name)
+        else:
+            self.codeuri = self._extract_code_uri(function_name, resource_properties, function_type)
+
+    def __enter__(self):
+        """ Dump Inline Code, if present for this Code Provider """
+        if self.inlinecode:
+            self._dump_code(self.inlinecode, self.codeuri, self.index)
+        return self
+
+    def __exit__(self, *exc):
+        """ CleanUp Resources created when using Inline code """
+        if self.inlinecode:
+            self._cleanup(self.codeuri, self.index)
+
+    def __repr__(self):
+        return repr(self.codeuri)
+
+    def __fspath__(self):
+        return self.codeuri
+
+    @staticmethod
+    def _setup_postfix(runtime):
+        index = 'index'
+        if runtime and runtime.startswith('python'):
+            return index + '.py'
+        return index
+
+    @staticmethod
+    def _extract_inline_code(resource_properties, function_type):
+        """ Parse InlineCode from and return the content or None """
+        code = None
+        if function_type == SamFunctionCodeProvider._SERVERLESS_FUNCTION:
```

Comment:
can we encapsulate this in a dictionary rather than multiple if else statements

Reply:
@TheSriram  mind you elaborate? I've basically kept consistent behavior based on `_extract_functions` function in samcli/commands/local/lib/sam_function_provider.py
```suggestion
            def _extract_functions(resources):
        """
        Extracts and returns function information from the given dictionary of SAM/CloudFormation resources. This
        method supports functions defined with AWS::Serverless::Function and AWS::Lambda::Function

        :param dict resources: Dictionary of SAM/CloudFormation resources
        :return dict(string : samcli.commands.local.lib.provider.Function): Dictionary of function LogicalId to the
            Function configuration object
        """

        result = {}

        for name, resource in resources.items():

            resource_type = resource.get("Type")
            resource_properties = resource.get("Properties", {})

            if resource_type == SamFunctionProvider._SERVERLESS_FUNCTION:
                layers = SamFunctionProvider._parse_layer_info(resource_properties.get("Layers", []), resources)
                codeuri = SamFunctionProvider._parse_function_code(name, resource_properties, resource_type)
                result[name] = SamFunctionProvider._convert_sam_function_resource(
                    name, resource_properties, layers, codeuri
                )

            elif resource_type == SamFunctionProvider._LAMBDA_FUNCTION:
                layers = SamFunctionProvider._parse_layer_info(resource_properties.get("Layers", []), resources)
                codeuri = SamFunctionProvider._parse_function_code(name, resource_properties, resource_type)
                result[name] = SamFunctionProvider._convert_lambda_function_resource(
                    name, resource_properties, layers, codeuri
                )

            # We don't care about other resource types. Just ignore them

        return result
```